### PR TITLE
fix(vpc): wait handler for public gateway and gateway network

### PR DESCRIPTION
**Pull request by Monitob**

### CODE REVIEW

File: scaleway/resource_vpc_public_gateway_ip.go [hunk 1]
```diff
@@ -153,10 +154,19 @@ func resourceScalewayVPCPublicGatewayIPDelete(ctx context.Context, d *schema.Res
...
-		Zone: zone,
-	}, scw.WithContext(ctx))
+	err = retryVPCContext(ctx, retryGWTimeout, func() *resource.RetryError {
+		errDeleteIP := vpcgwAPI.DeleteIP(&vpcgw.DeleteIPRequest{
+			IPID: ID,
+			Zone: zone,
+		}, scw.WithContext(ctx))
+		if errDeleteIP != nil {
+			if is409Error(errDeleteIP) || is412Error(errDeleteIP) {
+				return resource.RetryableError(errDeleteIP)
```
Comment (remyleone) [inline scaleway/resource_vpc_public_gateway_ip.go:164]: Can this go into an infinite loop?

File: scaleway/resource_vpc_private_network.go [hunk 2]
```diff
@@ -131,17 +142,20 @@ func resourceScalewayVPCPrivateNetworkDelete(ctx context.Context, d *schema.Reso
...
-			l.Warningf("error deleting private network in zone (%s): %s", zone, err)
-			return nil
+	err = retryVPCContext(ctx, retryGWTimeout, func() *resource.RetryError {
+		errPN := vpcAPI.DeletePrivateNetwork(&vpc.DeletePrivateNetworkRequest{
+			PrivateNetworkID: ID,
+			Zone:             zone,
+		}, scw.WithContext(ctx))
+		if errPN != nil {
+			if is409Error(errPN) || is412Error(errPN) {
+				return resource.RetryableError(errPN)
```
Comment (remyleone) [inline scaleway/resource_vpc_private_network.go:152]: Can this go into an infinite loop?

File: scaleway/helpers_vpc.go [hunk 3]
```diff
@@ -40,3 +44,60 @@ func vpcAPI(m interface{}) (*vpc.API, error) {
...
+		PollInterval: PollVPCTimeOut,
+		Refresh: func() (interface{}, string, error) {
+			rErr := f()
+
+			resultErrMu.Lock()
+			defer resultErrMu.Unlock()
+
+			if rErr == nil {
+				resultErr = nil
+				return 42, "success", nil
```
Comment (remyleone) [inline scaleway/helpers_vpc.go:76]: Why 42? Could we have something meaningful?

File: scaleway/helpers_vpc.go [hunk 4]
```diff
@@ -40,3 +44,60 @@ func vpcAPI(m interface{}) (*vpc.API, error) {
 
 	return vpc.NewAPI(meta.scwClient), nil
 }
+
+const minVPCTimeOut = 10 * time.Second
+const PollVPCTimeOut = 30 * time.Second
+
+// retryVPCContext is a basic wrapper around StateChangeConf that will just retry with a poll interval
```
Comment (remyleone) [inline scaleway/helpers_vpc.go:51]: Why is this change needed? Why can't we use a wait function around resources as we do in any other resource?

File: scaleway/resource_vpc_gateway_network.go [hunk 5]
```diff
@@ -189,23 +227,42 @@ func resourceScalewayVPCGatewayNetworkRead(ctx context.Context, d *schema.Resour
 		cleanUpDHCPValue = *expandBoolPtr(cleanUpDHCP)
 	}
 
-	_ = d.Set("gateway_id", newZonedID(zone, gatewayNetwork.GatewayID).String())
-	_ = d.Set("private_network_id", newZonedID(zone, gatewayNetwork.PrivateNetworkID).String())
-	_ = d.Set("enable_masquerade", gatewayNetwork.EnableMasquerade)
+	_ = d.Set("gateway_id", newZonedID(zone, gwn.GatewayID).String())
```
Comment (remyleone) [inline scaleway/resource_vpc_gateway_network.go:230]: Why is this change needed? Using a full variable instead of an acronym is better and would make the diff smaller.

File: scaleway/resource_vpc_gateway_network.go [hunk 6]
```diff
@@ -189,23 +227,42 @@ func resourceScalewayVPCGatewayNetworkRead(ctx context.Context, d *schema.Resour
...
-	_ = d.Set("created_at", gatewayNetwork.CreatedAt.Format(time.RFC3339))
-	_ = d.Set("updated_at", gatewayNetwork.UpdatedAt.Format(time.RFC3339))
+	_ = d.Set("created_at", gwn.CreatedAt.Format(time.RFC3339))
+	_ = d.Set("updated_at", gwn.UpdatedAt.Format(time.RFC3339))
 	_ = d.Set("zone", zone.String())
 
 	return nil
 }
 
+func gatewayNetworkAttributeIsMissing(gwn *vpcgw.GatewayNetwork) bool {
```
Comment (remyleone) [inline scaleway/resource_vpc_gateway_network.go:241]: Could the name be `isGatewayNetworkMacAddressEmpty`?

File: scaleway/resource_vpc_private_network.go (hunk 2 above)
Comment (Monitob) [reply scaleway/resource_vpc_private_network.go:152]: The loop is limited by a timeout and the client allow 4  attempts

File: scaleway/helpers_vpc.go (hunk 3 above)
Comment (Monitob) [reply scaleway/helpers_vpc.go:76]: I've took the inspiration from the terraform helpers.  The difference here that we can handle the poll intervals as the timeouts parameters.

File: scaleway/resource_vpc_public_gateway_ip.go (hunk 1 above)
Comment (Monitob) [reply scaleway/resource_vpc_public_gateway_ip.go:164]: Yes, It is recommended in this case. This resource could been used by another resource and until today we don't have a way to verify before deleted

File: scaleway/helpers_vpc.go (hunk 4 above)
Comment (Monitob) [reply scaleway/helpers_vpc.go:51]: We can ask a resource status in this case.

File: scaleway/resource_vpc_gateway_network.go (hunk 5 above)
Comment (Monitob) [reply scaleway/resource_vpc_gateway_network.go:230]: Ok, It will be change in next commit

File: scaleway/resource_vpc_gateway_network.go [hunk 7]
```diff
@@ -243,7 +300,7 @@ func resourceScalewayVPCGatewayNetworkDelete(ctx context.Context, d *schema.Reso
 		Zone:             zone,
 		RetryInterval:    &defaultInterval,
 	})
-	if err != nil && !is404Error(err) {
+	if err != nil {
```
Comment (remyleone) [inline scaleway/resource_vpc_gateway_network.go:303]: Why do we ignore this case now?

File: scaleway/resource_vpc_gateway_network.go [hunk 8]
```diff
@@ -124,18 +127,29 @@ func resourceScalewayVPCGatewayNetworkCreate(ctx context.Context, d *schema.Reso
 		Timeout:       scw.TimeDurationPtr(gatewayWaitForTimeout),
 		RetryInterval: &retryInterval,
 	}, scw.WithContext(ctx))
-
-	if err != nil && !is404Error(err) {
+	if err != nil {
```
Comment (remyleone) [inline scaleway/resource_vpc_gateway_network.go:130]: Why do we ignore this case now?

File: scaleway/resource_vpc_gateway_network.go [hunk 9]
```diff
@@ -155,15 +169,39 @@ func resourceScalewayVPCGatewayNetworkRead(ctx context.Context, d *schema.Resour
...
-			d.SetId("")
-			return nil
+		return diag.FromErr(err)
+	}
+
+	var gatewayNetwork *vpcgw.GatewayNetwork
+	err = retryVPCContext(ctx, retryGWTimeout, func() *resource.RetryError {
+		currentGW, errReadVPC := vpcgwNetworkAPI.GetGatewayNetwork(readGWNetwork, scw.WithContext(ctx))
+		if isGatewayNetworkMacAddressEmpty(currentGW) {
+			return resource.RetryableError(fmt.Errorf("missing attribute"))
```
Comment (remyleone) [inline scaleway/resource_vpc_gateway_network.go:193]: This message is not clear. Could you put something more explicit about what is going on? For instance: Gateway Network got no mac addresses.

File: scaleway/resource_vpc_gateway_network.go (hunk 8 above)
Comment (Monitob) [reply scaleway/resource_vpc_gateway_network.go:127]: We need explicitly that the gateway exist.

File: scaleway/testdata/k8s-cluster-pool-wait.cassette.yaml [hunk 10]
```diff
@@ -6,18 +6,25 @@ interactions:
     form: {}
```
Comment (remyleone) [inline scaleway/testdata/k8s-cluster-pool-wait.cassette.yaml]: Why is this file modified?

File: scaleway/testdata/k8s-cluster-pool-zone.cassette.yaml [hunk 11]
```diff
@@ -6,18 +6,25 @@ interactions:
     form: {}
```
[the remark this answers is not in the document]
Comment (Monitob) [reply scaleway/testdata/k8s-cluster-pool-zone.cassette.yaml]: I've removed this commit.

[the remark this answers is not in the document]
Comment (remyleone) [reply scaleway/testdata/k8s-cluster-pool-zone.cassette.yaml]: There is still k8s file :(

File: scaleway/helpers_vpcgw.go [hunk 12]
```diff
@@ -36,3 +37,33 @@ func vpcgwAPIWithZoneAndID(m interface{}, id string) (*vpcgw.API, scw.Zone, stri
...
+	api  *vpcgw.API
+	zone scw.Zone
+}
+
+func newVPCHandler(ctx context.Context, api *vpcgw.API, zone scw.Zone) *vpcHandler {
+	return &vpcHandler{ctx: ctx, api: api, zone: zone}
+}
+
+func (vpcH *vpcHandler) waitGNetwork(gwNetworkID string) (*vpcgw.GatewayNetwork, error) {
+	retryInterval := retryIntervalVPCGatewayNetwork
```
Comment (remyleone) [inline scaleway/helpers_vpcgw.go:52]: Have this code directly inline. We need to stay consistent with the rest of the wait operation across the provider.

File: scaleway/testdata/data-source-k8s-pool-basic.cassette.yaml [hunk 13]
```diff
@@ -1671,10 +1671,10 @@ interactions:
     url: https://api.scaleway.com/k8s/v1/regions/fr-par/clusters/7ffe4ef9-6fd5-4db2-b7e1-55b65c0b77be/nodes?order_by=created_at_asc&page=1&pool_id=a0698de1-7396-471c-959a-8a10f2075e00&status=unknown
```
Comment (remyleone) [inline scaleway/testdata/data-source-k8s-pool-basic.cassette.yaml:1671]: k8s cassettes

File: scaleway/testdata/k8s-cluster-pool-basic.cassette.yaml [hunk 14]
```diff
@@ -119,7 +119,7 @@ interactions:
...05bdce1-64c0-48ab-899d-868455867ecf","created_at":"2021-12-15T13:34:57.126638Z","updated_at":"2021-12-15T13:35:00.514930Z","type":"kapsule","name":"K8SPoolConfigMinimal","description":"","status":"pool_required","version":"1.23.0","cni":"calico","tags":["terraform-test","scaleway_k8s_cluster","minimal"],"cluster_url":"https://41e0ee52-5f80-4772-bf6f-6b20411133de.api.k8s.fr-par.scw.cloud:6443","dns_wildcard":"*.41e0ee52-5f80-4772-bf6f-6b20411133de.nodes.k8s.fr-par.scw.cloud","autoscaler_config":{"scale_down_disabled":false,"scale_down_delay_after_add":"10m","estimator":"binpacking","expander":"random","ignore_daemonsets_utilization":false,"balance_similar_node_groups":false,"expendable_pods_priority_cutoff":0,"scale_down_unneeded_time":"10m","scale_down_utilization_threshold":0.5,"max_graceful_termination_sec":0},"dashboard_enabled":false,"ingress":"none","auto_upgrade":{"enabled":false,"maintenance_window":{"start_hour":0,"day":"any"}},"upgrade_available":false,"feature_gates":[],"admission_plugins":[],"open_id_connect_config":{"issuer_url":"","client_id":"","username_claim":"","username_prefix":"","groups_claim":[],"groups_prefix":"","required_claim":[]},"apiserver_cert_sans":[]}'
```
Comment (remyleone) [inline scaleway/testdata/k8s-cluster-pool-basic.cassette.yaml:119]: k8s cassettes

File: scaleway/testdata/k8s-cluster-pool-placement-group.cassette.yaml [hunk 15]
```diff
@@ -923,7 +923,7 @@ interactions:
..."105bdce1-64c0-48ab-899d-868455867ecf","created_at":"2021-12-15T13:34:57.100044Z","updated_at":"2021-12-15T13:35:05.526646Z","type":"kapsule","name":"placement_group","description":"","status":"creating","version":"1.23.0","cni":"calico","tags":["terraform-test","scaleway_k8s_cluster","placement_group"],"cluster_url":"https://c849e052-2f76-4b2e-a016-f20eb523321f.api.k8s.fr-par.scw.cloud:6443","dns_wildcard":"*.c849e052-2f76-4b2e-a016-f20eb523321f.nodes.k8s.fr-par.scw.cloud","autoscaler_config":{"scale_down_disabled":false,"scale_down_delay_after_add":"10m","estimator":"binpacking","expander":"random","ignore_daemonsets_utilization":false,"balance_similar_node_groups":false,"expendable_pods_priority_cutoff":0,"scale_down_unneeded_time":"10m","scale_down_utilization_threshold":0.5,"max_graceful_termination_sec":0},"dashboard_enabled":false,"ingress":"none","auto_upgrade":{"enabled":false,"maintenance_window":{"start_hour":0,"day":"any"}},"upgrade_available":false,"feature_gates":[],"admission_plugins":[],"open_id_connect_config":{"issuer_url":"","client_id":"","username_claim":"","username_prefix":"","groups_claim":[],"groups_prefix":"","required_claim":[]},"apiserver_cert_sans":[]}'
```
Comment (remyleone) [inline scaleway/testdata/k8s-cluster-pool-placement-group.cassette.yaml:923]: k8s cassettes

File: scaleway/testdata/k8s-cluster-pool-wait.cassette.yaml [hunk 16]
```diff
@@ -5207,7 +5207,7 @@ interactions:
     body: '{"region":"fr-par","id":"13d550de-9d49-4036-bb60-6b04a3130653","cluster_id":"e2ce45e7-fe8a-4807-b418-81795615d803","created_at":"2021-12-15T13:40:04.444980Z","updated_at":"2021-12-15T13:45:56.114009Z","name":"minimal","status":"ready","version":"1.23.0","node_type":"gp1_xs","autoscaling":false,"size":2,"min_size":1,"max_size":2,"container_runtime":"containerd","autohealing":false,"tags":[],"placement_group_id":null,"kubelet_args":{},"upgrade_policy":{"max_unavailable":1,"max_surge":0},"zone":"fr-par-1"}'
```
Comment (remyleone) [inline scaleway/testdata/k8s-cluster-pool-wait.cassette.yaml:5207]: k8s cassettes

File: scaleway/helpers_vpcgw.go (hunk 12 above)
Comment (remyleone) [reply scaleway/helpers_vpcgw.go:52]: We will tackle all the issues related to waiting in a single PR in the future that will add consistent support across all resources. This PR will also handle retry and timeout.

File: scaleway/testdata/data-source-k8s-pool-basic.cassette.yaml (hunk 13 above)
Comment (Monitob) [reply scaleway/testdata/data-source-k8s-pool-basic.cassette.yaml]: Thanks,  I've just fixed this

File: scaleway/helpers_vpcgw.go (hunk 12 above)
Comment (Monitob) [reply scaleway/helpers_vpcgw.go]: Updated on last commit

**Review by remyleone — APPROVED** [review]
